File: core/src/main/java/vn/loitp/core/utilities/LStoreUtil.java
package vn.loitp.core.utilities;

import android.app.Activity;
import android.content.Context;
import android.graphics.Color;
import android.os.AsyncTask;
import android.os.Environment;
import android.os.StatFs;

import java.io.BufferedReader;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.URL;
import java.util.Random;

public class LStoreUtil {
    static String TAG = LStoreUtil.class.getSimpleName();
    static String folderName = "Loitp";
    static String folderPath;

    public static String FOLDER_TRANSLATE = ".Loitp";
    public static String FILE_TRANSLATE_FAV_SENTENCE = "Loitp.txt";

    private static final String EXTENSION = ".db";
    public static final String FOLDER_TRUYENTRANHTUAN = "foldertruyentranhtuan";
    public static final String FILE_NAME_MAIN_COMICS_LIST_HTML_CODE = "filenamemaincomicslisthtmlcode" + EXTENSION;
    public static final String FILE_NAME_MAIN_COMICS_LIST = "filenamemaincomicslist" + EXTENSION;
    public static final String FILE_NAME_MAIN_COMICS_LIST_FAVOURITE = "filenamemaincomicslistfavourite" + EXTENSION;
    public static final String FILE_NAME_TRUYENTRANHTUAN_DOWNLOADED_COMIC = "filenamedownloadedcomic" + EXTENSION;

    public static String getFileNameComic(String url) {
        url = url.replace("/", "");
        url = url.replace(".", "");
        url = url.replace(":", "");
        url = url.replace("-", "");
        return url + EXTENSION;
    }

    public static String createFileImage(int i) {
        return "p" + i + EXTENSION;
    }

    public interface CallbackReadFile {
        public void onFinish(String result);
    }

    public interface CallbackWriteFile {
        public void onFinish(boolean isSuccess);
    }

    public static String getFolderPath(Context context) {
        if (isSdPresent()) {
            try {
                File sdPath = new File(Environment.getExternalStorageDirectory().getAbsolutePath() + "/" + folderName);
                if (!sdPath.exists()) {
                    sdPath.mkdirs();
                    folderPath = sdPath.getAbsolutePath();
                } else if (sdPath.exists()) {
                    folderPath = sdPath.getAbsolutePath();
                }
            } catch (Exception e) {
                LLog.d("TAG", "if getFolderPath: " + e.toString());
            }
            folderPath = Environment.getExternalStorageDirectory().getPath() + "/" + folderName + "/";
        } else {
            try {
                File cacheDir = new File(context.getCacheDir(), folderName + "/");
                if (!cacheDir.exists()) {
                    cacheDir.mkdirs();
                    folderPath = cacheDir.getAbsolutePath();
                } else if (cacheDir.exists()) {
                    folderPath = cacheDir.getAbsolutePath();
                }
            } catch (Exception e) {
                LLog.d("TAG", "else getFolderPath: " + e.toString());
            }
        }
        return folderPath;
    }

    public static boolean isSdPresent() {
        return Environment.getExternalStorageState().equals(Environment.MEDIA_MOUNTED);
    }

    public static int checkSDCardFreeSize() {
        StatFs stat = new StatFs(Environment.getExternalStorageDirectory().getPath());
        long bytesAvailable = (long) stat.getBlockSize() * (long) stat.getAvailableBlocks();
        long megAvailable = bytesAvailable / (1024 * 1024);
        return (int) megAvailable;
    }

    /*
    save tring json to sdcard
    ex: writeToFile("module.json", strJson);
     */
    public static boolean writeToFile(Activity activity, String folder, String fileName, String body) {
        boolean isCompelete = true;
        FileOutputStream fos = null;
        try {
            String path = LStoreUtil.getFolderPath(activity);
            if (folder != null) {
                //path = path + "/" + folder;
                //path = path + folder;
                path = path + folder + "/";
            }
            LLog.d(TAG, "path: " + path);
            final File dir = new File(path);
            if (!dir.exists()) {
                if (!dir.mkdirs()) {
                    LLog.d(TAG, "could not create the directories");
                }
            }
            final File myFile = new File(dir, fileName);
            if (!myFile.exists()) {
                boolean isSuccess = myFile.createNewFile();
                LLog.d(TAG, "isSuccess: " + isSuccess);
            }
            fos = new FileOutputStream(myFile);
            fos.write(body.getBytes());
            fos.close();
        } catch (IOException e) {
            LLog.d(TAG, e.toString());
            isCompelete = false;
        }
        return isCompelete;
    }

    public static void writeToFile(final Activity activity, final String folder, final String fileName, final String body, final CallbackWriteFile callbackWriteFile) {
        new AsyncTask<Void, Void, Void>() {
            boolean isSuccess;

            @Override
            protected Void doInBackground(Void... params) {
                isSuccess = writeToFile(activity, folder, fileName, body);
                return null;
            }

            @Override
            protected void onPostExecute(Void aVoid) {
                super.onPostExecute(aVoid);
                if (callbackWriteFile != null) {
                    callbackWriteFile.onFinish(isSuccess);
                }
            }
        }.execute();
    }

    /*
       * read text file from folder
       */
    public static String readTxtFromFolder(Activity activity, String folderName, String fileName) {
        String path = LStoreUtil.getFolderPath(activity) + (folderName == null ? "/" : (folderName + "/")) + fileName;
        LLog.d(TAG, "path: " + path);
        File txtFile = new File(path);
        StringBuilder text = new StringBuilder();
        try {
            BufferedReader reader = new BufferedReader(new FileReader(txtFile));
            String line;
            while ((line = reader.readLine()) != null) {
                text.append(line + '\n');
            }
            reader.close();
        } catch (IOException e) {
            LLog.d(TAG, "readTxtFromFolder===" + e.toString());
        }
        return text.toString();
    }

    public static void readTxtFromFolder(final Activity activity, final String folderName, final String fileName, final CallbackReadFile callbackReadFile) {
        new AsyncTask<Void, Void, Void>() {
            String result = "";

            @Override
            protected Void doInBackground(Void... params) {
                result = readTxtFromFolder(activity, folderName, fileName);
                return null;
            }

            @Override
            protected void onPostExecute(Void aVoid) {
                super.onPostExecute(aVoid);
                callbackReadFile.onFinish(result);
            }
        }.execute();
    }

    public interface EventReadFromFolder {
        public void onSuccess(String data);

        public void onError();
    }

    /*
   * read text file from folder in background
   */
    public static void readTxtFromFolder(final Activity activity, final String folderName, final String fileName, final EventReadFromFolder eventReadFromFolder) {
        new AsyncTask<Void, Void, Void>() {
            private StringBuilder text = null;
            private boolean runTaskSuccess = true;

            @Override
            protected void onPreExecute() {
                super.onPreExecute();
            }

            @Override
            protected Void doInBackground(Void... params) {
                String path = LStoreUtil.getFolderPath(activity) + (folderName == null ? "/" :
                        (folderName + "/")) + fileName;
                LLog.d(TAG, "path: " + path);
                File txtFile = new File(path);
                text = new StringBuilder();
                try {
                    BufferedReader reader = new BufferedReader(new FileReader(txtFile));
                    String line;
                    while ((line = reader.readLine()) != null) {
                        text.append(line + '\n');
                    }
                    reader.close();
                } catch (IOException e) {
                    runTaskSuccess = false;
                    LLog.d(TAG, "readTxtFromFolder===" + e.toString());

                }
                return null;
            }

            @Override
            protected void onPostExecute(Void aVoid) {
                if (runTaskSuccess) {
                    eventReadFromFolder.onSuccess(text.toString());
                } else {
                    eventReadFromFolder.onError();
                }
                super.onPostExecute(aVoid);
            }
        }.execute();
    }

    /*
    * read text file in raw folder
    */
    public static String readTxtFromRawFolder(Context context, int nameOfRawFile) {
        InputStream inputStream = context.getResources().openRawResource(nameOfRawFile);
        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
        int i;
        try {
            i = inputStream.read();
            while (i != -1) {
                byteArrayOutputStream.write(i);
                i = inputStream.read();
            }
            inputStream.close();
        } catch (Exception e) {
            LLog.d(TAG, e.toString());
        }
        return byteArrayOutputStream.toString();
    }

    public interface CallbackReadFromRaw {
        public void onFinish(String result);
    }

    public static void readTxtFromRawFolder(final Context context, final int nameOfRawFile, final CallbackReadFromRaw callbackReadFromRaw) {
        new AsyncTask<Void, Void, Void>() {
            String result;

            @Override
            protected Void doInBackground(Void... params) {
                result = readTxtFromRawFolder(context, nameOfRawFile);
                return null;
            }

            @Override
            protected void onPostExecute(Void aVoid) {
                super.onPostExecute(aVoid);
                callbackReadFromRaw.onFinish(result);
            }
        }.execute();
    }

    /*@param link: the url of the website
      * return true if save success
      * return false if save failed
      * */
    /*public static boolean saveHTMLCodeFromURLToSDCard(Context context, String link, String
    folderName, String fileName) {
        return saveHTMLCodeFromURLToSDCard((Activity) context, link, folderName, fileName);
    }*/

    public static boolean saveHTMLCodeFromURLToSDCard(Context context, String link, String folderName, String fileName) {
        boolean state = false;
        InputStream is = null;
        try {
            URL url = new URL(link);
            is = url.openStream();
            BufferedReader br = new BufferedReader(new InputStreamReader(is));
            String line;
            StringBuilder stringBuilder = new StringBuilder();
            while ((line = br.readLine()) != null) {
                stringBuilder.append(line);
            }
            br.close();
            is.close();
            //LLog.d(TAG, "saveHTMLCodeFromURLToSDCard success: " + stringBuilder.toString());
            writeToFile((Activity) context, folderName, fileName, stringBuilder.toString());
            state = true;
        } catch (Exception e) {
            //LLog.d(TAG, "saveHTMLCodeFromURLToSDCard failed: " + e.toString());
        }
        return state;
    }

    /*
      save bitmap to sdcard
      @param bitmap
      @param subDirName like Quote
       */
    /*public void saveBitmapToSDCard(Bitmap bitmap, String subDirName) {
        //String root = Environment.getExternalStorageDirectory().toString();
        //File myDir = new File(root + "/saved_images");
        String root = LStoreUtil.getFolderPath(mContext);
        File myDir = new File(root + "/" + subDirName);
        if (!myDir.exists())
            myDir.mkdirs();
        Random generator = new Random();
        int n = 10000;
        n = generator.nextInt(n);
        String fname = "Image-" + n + ".jpg";
        File file = new File(myDir, fname);
        if (file.exists()) file.delete();
        try {
            FileOutputStream out = new FileOutputStream(file);
            bitmap.compress(Bitmap.CompressFormat.JPEG, 90, out);
            out.flush();
            out.close();
            //LToast.dialog(mContext, R.string.down_ok, 0);
        } catch (Exception e) {
            LLog.dialog(TAG, "saveBitmapToSDCard: " + e.toString());
        }
    }*/

    /*
       * get random quote
       */
    /*public String getRandomeQuote() {
        InputStream in;
        // quote
        String[] arr_Str = null;
        String str = "";
        try {
            in = mContext.getAssets().open("quote.txt");
            byte[] buffer = new byte[in.available()];
            in.read(buffer);
            in.close();
            String chuoi = new String(buffer);
            arr_Str = chuoi.split("###");

            // for (int i = 0; i < arr_Str.length; i++) {
            // // 1 ki tu dau tien cua moi chap
            // String char1;
            // String[] arrChar = arr_Str[i].split(" ");
            // try {
            // char1 = arrChar[1];
            // } catch (Exception e) {
            // char1 = arr_Str[i].substring(0, 1);
            // }
            // }

            // Random r = new Random();
            // int ran = r.nextInt(arr_Str.length);
            int ran = getRandomNumber(arr_Str.length);
            str = arr_Str[ran];

        } catch (Exception e) {
            // LLog.dialog("bug", e.toString());
        }
        return str;
    }*/

    /*
     * get random quote
	 */
    /*public static String getRandomeQuote(Context context) {
        InputStream in;
        String[] arr_Str = null;
        String str = "";
        try {
            in = context.getAssets().open("raw/quote.txt");
            byte[] buffer = new byte[in.available()];
            in.read(buffer);
            in.close();
            String chuoi = new String(buffer);
            arr_Str = chuoi.split("###");
            int ran = getRandomNumber(arr_Str.length);
            str = arr_Str[ran];

        } catch (Exception e) {
            LLog.d(TAG, "getRandomeQuote: " + e.toString());
        }
        return str;
    }*/

    /*
     * get random number
     */
    public static int getRandomNumber(int length) {
        int x = 0;
        Random r = new Random();
        x = r.nextInt(length);
        return x;
    }

    public static int getRandomColor() {
        Random rnd = new Random();
        return Color.argb(255, rnd.nextInt(256), rnd.nextInt(256), rnd.nextInt(256));
    }

    public static String getPathOfFileNameMainComicsListHTMLCode(Context context) {
        return getFolderPath(context) + "/" + LStoreUtil.FOLDER_TRUYENTRANHTUAN + "/" + LStoreUtil.FILE_NAME_MAIN_COMICS_LIST_HTML_CODE;
    }
}

File: core/src/main/java/com/core/utilities/LStoreUtil.kt
package com.core.utilities

import alirezat775.lib.downloader.Downloader
import alirezat775.lib.downloader.core.OnDownloadListener
import android.app.ActivityManager
import android.content.Context
import android.graphics.Color
import android.media.MediaScannerConnection
import android.os.Environment
import com.core.base.BaseApplication
import com.google.gson.reflect.TypeToken
import com.model.App
import com.model.GG
import okhttp3.*
import java.io.*
import java.util.*
import java.util.zip.ZipEntry
import java.util.zip.ZipInputStream
import kotlin.collections.ArrayList

//https://gist.github.com/lopspower/76421751b21594c69eb2
//https://github.com/lopspower/BestAndroidGists

class LStoreUtil {
    companion object {
        internal var logTag = LStoreUtil::class.java.simpleName

        const val FOLDER_TRANSLATE = ".Loitp"
        const val FILE_TRANSLATE_FAV_SENTENCE = "Loitp.txt"
        private const val EXTENSION = ".txt"
        const val FOLDER_TRUYENTRANHTUAN = "zloitpbestcomic"
        const val FILE_NAME_MAIN_COMICS_LIST_HTML_CODE = "filenamemaincomicslisthtmlcode$EXTENSION"
        const val FILE_NAME_MAIN_COMICS_LIST = "filenamemaincomicslist$EXTENSION"
        const val FILE_NAME_MAIN_COMICS_LIST_FAVOURITE = "filenamemaincomicslistfavourite$EXTENSION"
        const val FILE_NAME_TRUYENTRANHTUAN_DOWNLOADED_COMIC = "filenamedownloadedcomic$EXTENSION"

        val isSdPresent: Boolean
            get() = Environment.getExternalStorageState() == Environment.MEDIA_MOUNTED

        val randomColor: Int
            get() {
                val rnd = Random()
                return Color.argb(255, rnd.nextInt(256), rnd.nextInt(256), rnd.nextInt(256))
            }

        //https://stackoverflow.com/questions/30703652/how-to-generate-light-and-pastel-colors-randomly-in-android
        // This is the base color which will be mixed with the generated one
        val randomColorLight: Int
            get() {
                val mRandom = Random(System.currentTimeMillis())
                val baseColor = Color.WHITE

                val baseRed = Color.red(baseColor)
                val baseGreen = Color.green(baseColor)
                val baseBlue = Color.blue(baseColor)

                val red = (baseRed + mRandom.nextInt(256)) / 2
                val green = (baseGreen + mRandom.nextInt(256)) / 2
                val blue = (baseBlue + mRandom.nextInt(256)) / 2

                return Color.rgb(red, green, blue)
            }

        val colors: IntArray
            get() = intArrayOf(Color.parseColor("#1BFFFF"), Color.parseColor("#2E3192"), Color.parseColor("#ED1E79"), Color.parseColor("#009E00"), Color.parseColor("#FBB03B"), Color.parseColor("#D4145A"), Color.parseColor("#3AA17E"), Color.parseColor("#00537E"))

        val texts: Array<String>
            get() = arrayOf("Relax, its only ONES and ZEROS !", "Hardware: The parts of a computer system that can be kicked.", "Computer dating is fine, if you're a computer.", "Better to be a geek than an idiot.", "If you don't want to be replaced by a computer, don't act like one.", "I'm not anti-social; I'm just not user friendly", "Those who can't write programs, write help files.", "The more I C, the less I see.  ")

        fun getFileNameComic(
                url: String
        ): String {
            var u = url
            u = u.replace(oldValue = "/", newValue = "")
            u = u.replace(oldValue = ".", newValue = "")
            u = u.replace(oldValue = ":", newValue = "")
            u = u.replace(oldValue = "-", newValue = "")
            return u + EXTENSION
        }

        fun createFileImage(
                i: Int
        ): String {
            return "p$i$EXTENSION"
        }

        //dung de bao hieu cho gallery load lai photo vi co anh moi
        fun sendBroadcastMediaScan(
                file: File? = null
        ) {
            file?.let {
//                val mediaScanIntent = Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE)
//                val contentUri = Uri.fromFile(file)
//                mediaScanIntent.data = contentUri
//                LAppResource.application.sendBroadcast(mediaScanIntent)

                MediaScannerConnection.scanFile(LAppResource.application,
                        arrayOf(file.toString()),
                        arrayOf(it.name),
                        null)
            }
        }

        @JvmOverloads
        fun getFolderPath(
                folderName: String = "z1000"
        ): String {
            var folderPath = ""
            if (isSdPresent) {
                try {
//                    C1
//                    val file = File(Environment.getExternalStorageDirectory().absolutePath + "/" + folderName)
//                        ex: /storage/emulated/0/ZZZTestDownloader

//                    C2
//                    val file = File(LAppResource.application.getExternalFilesDir(null)?.absolutePath + "/" + folderName)
//                    ex: /storage/emulated/0/Android/data/loitp.basemaster/files/ZZZTestDownloader

//                    C3
                    val path = LAppResource.application.getExternalFilesDir(null)?.parent?.split("/Andro")?.get(0)
                            ?: ""
                    val file = File("$path/$folderName")
                    LLog.d(logTag, "file ${file.path}")
                    LLog.d(logTag, "file exists " + file.exists())

                    folderPath = if (file.exists()) {
                        file.absolutePath
                    } else {
                        file.mkdirs()
                        file.absolutePath
                    }

                } catch (e: Exception) {
                    e.printStackTrace()
                    LLog.e(logTag, "err isSdPresent $e")
                }
            } else {
                try {
                    val cacheDir = File(LAppResource.application.cacheDir, "$folderName/")
                    if (!cacheDir.exists()) {
                        cacheDir.mkdirs()
                        folderPath = cacheDir.absolutePath
                    } else if (cacheDir.exists()) {
                        folderPath = cacheDir.absolutePath
                    }
                } catch (e: Exception) {
                    e.printStackTrace()
                    LLog.e(logTag, "err !isSdPresent $e")
                }

            }
            LLog.d(logTag, "return getFolderPath folderPath $folderPath")
            return folderPath
        }

        /*
        save string json to sdcard
         */
        fun writeToFile(
                folder: String?,
                fileName: String,
                body: String
        ): File? {
            val fos: FileOutputStream?
            try {
                var path = getFolderPath()
                if (folder != null) {
                    path = "$path/$folder/"
                }
                val dir = File(path)
                val dirExist = dir.exists()
                if (!dirExist) {
                    if (!dir.mkdirs()) {
                        return null
                    }
                }
                val myFile = File(dir, fileName)
                val myFileExist = myFile.exists()
                if (!myFileExist) {
                    val isSuccess = myFile.createNewFile()
                }
                fos = FileOutputStream(myFile)
                fos.write(body.toByteArray())
                fos.close()
                LLog.d(logTag, "<<<writeToFile myFile path: " + myFile.path)
                return myFile
            } catch (e: IOException) {
                e.printStackTrace()
                return null
            }
        }

        /*
         * read text file from folder
         */
        fun readTxtFromFolder(
                folderName: String?,
                fileName: String
        ): String {
            val path = getFolderPath() + (if (folderName == null) "/" else "/$folderName/") + fileName
            val txtFile = File(path)
            LLog.d(logTag, "readTxtFromFolder txtFile ${txtFile.path}")
            val text = StringBuilder()
            try {
                val reader = BufferedReader(FileReader(txtFile))
                var line: String? = null
                while ({ line = reader.readLine(); line }() != null) {
                    text.append(line + '\n')
                }
                reader.close()
            } catch (e: IOException) {
                e.printStackTrace()
            }
            return text.toString()
        }

        /*
         * read text file in raw folder
         */
        fun readTxtFromRawFolder(
                nameOfRawFile: Int
        ): String {
            val inputStream = LAppResource.application.resources.openRawResource(nameOfRawFile)
            val byteArrayOutputStream = ByteArrayOutputStream()
            var i: Int
            try {
                i = inputStream.read()
                while (i != -1) {
                    byteArrayOutputStream.write(i)
                    i = inputStream.read()
                }
                inputStream.close()
            } catch (e: Exception) {
                e.printStackTrace()
            }

            return byteArrayOutputStream.toString()
        }

        fun readTxtFromAsset(
                assetFile: String
        ): String {
            val ins: InputStream
            var str = ""
            try {
                ins = LAppResource.application.assets.open(assetFile)
                val buffer = ByteArray(ins.available())
                ins.read(buffer)
                ins.close()
                str = String(buffer)

            } catch (e: Exception) {
                e.printStackTrace()
            }

            return str
        }

        /*
         * get random number
         */
        fun getRandomNumber(
                length: Int
        ): Int {
            val r = Random()
            return r.nextInt(length)
        }

        fun getPathOfFileNameMainComicsListHTMLCode(): String {
            return getFolderPath() + FOLDER_TRUYENTRANHTUAN + "/" + FILE_NAME_MAIN_COMICS_LIST_HTML_CODE
        }

        fun getFileFromAssets(
                fileName: String
        ): File? {

            val file = File(LAppResource.application.cacheDir.toString() + "/" + fileName)
            if (!file.exists())
                try {
                    val ins = LAppResource.application.assets.open(fileName)
                    val size = ins.available()
                    val buffer = ByteArray(size)
                    ins.read(buffer)
                    ins.close()
                    val fos = FileOutputStream(file)
                    fos.write(buffer)
                    fos.close()
                } catch (e: Exception) {
                    e.printStackTrace()
                    return null
                }

            return file
        }

        fun getListEpubFiles(
                parentDir: File
        ): ArrayList<File> {
            val listFile = ArrayList<File>()
            val files = parentDir.listFiles()
            if (files != null) {
                for (file in files) {
                    if (file.isDirectory) {
                        listFile.addAll(getListEpubFiles(file))
                    } else {
                        if (file.name.endsWith(".epub")) {
                            listFile.add(file)
                        }
                    }
                }
            }
            return listFile
        }

        fun getSettingFromGGDrive(
                linkGGDriveSetting: String? = null,
                onGGFailure: ((call: Call, e: IOException) -> Unit)? = null,
                onGGResponse: ((app: App?, isNeedToShowMsg: Boolean) -> Unit)? = null
        ) {
            if (linkGGDriveSetting == null || linkGGDriveSetting.isEmpty()) {
                return
            }
            val request = Request.Builder().url(linkGGDriveSetting).build()
            val okHttpClient = OkHttpClient()
            okHttpClient.newCall(request).enqueue(object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    onGGFailure?.invoke(call, e)
                }

                @Throws(IOException::class)
                override fun onResponse(call: Call, response: Response) {
                    if (response.isSuccessful) {
                        val responseBody = response.body ?: return
                        val json = responseBody.string()
                        val app = BaseApplication.gson.fromJson(json, App::class.java)
                        if (app == null) {
                            onGGResponse?.invoke(null, false)
                        } else {
                            val localMsg = LPrefUtil.getGGAppMsg()
                            val serverMsg = app.config?.msg
                            LPrefUtil.setGGAppMsg(serverMsg)
                            val isNeedToShowMsg: Boolean
                            isNeedToShowMsg = if (serverMsg.isNullOrEmpty()) {
                                false
                            } else {
                                !localMsg?.trim { it <= ' ' }.equals(serverMsg, ignoreCase = true)
                            }
                            onGGResponse?.invoke(app, isNeedToShowMsg)
                        }

                    } else {
                        onGGResponse?.invoke(null, false)
                    }
                }
            })
        }

        fun getTextFromGGDrive(
                linkGGDrive: String? = null,
                onGGFailure: ((call: Call, e: Exception) -> Unit)? = null,
                onGGResponse: ((listGG: ArrayList<GG>) -> Unit)? = null
        ) {
            if (linkGGDrive.isNullOrEmpty()) {
                return
            }
            val request = Request.Builder().url(linkGGDrive).build()
            val okHttpClient = OkHttpClient()
            okHttpClient.newCall(request).enqueue(object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    onGGFailure?.invoke(call, e)
                }

                @Throws(IOException::class)
                override fun onResponse(call: Call, response: Response) {
                    if (response.isSuccessful) {
                        val responseBody = response.body
                        val json = responseBody?.string()
                        if (json.isNullOrEmpty()) {
                            onGGFailure?.invoke(call, NullPointerException("responseBody isNullOrEmpty"))
                        } else {
                            val listGG: ArrayList<GG> = BaseApplication.gson.fromJson(json, object : TypeToken<List<GG?>?>() {}.type)
                            onGGResponse?.invoke(listGG)
                        }
                    } else {
                        onGGFailure?.invoke(call, NullPointerException("responseBody !isSuccessful"))
                    }
                }
            })
        }

        @Suppress("INTEGER_OVERFLOW")
        fun getSize(
                size: Int
        ): String {
            var s = ""
            val kb = (size / 1024).toDouble()
            val mb = kb / 1024
            val gb = kb / 1024
            val tb = kb / 1024
            if (size < 1024) {
                s = "$size Bytes"
            } else if (size >= 1024 && size < 1024 * 1024) {
                s = String.format("%.2f", kb) + " KB"
            } else if (size >= 1024 * 1024 && size < 1024 * 1024 * 1024) {
                s = String.format("%.2f", mb) + " MB"
            } else if (size >= 1024 * 1024 * 1024 && size < 1024 * 1024 * 1024 * 1024) {
                s = String.format("%.2f", gb) + " GB"
            } else if (size >= 1024 * 1024 * 1024 * 1024) {
                s = String.format("%.2f", tb) + " TB"
            }
            return s
        }

        fun getAvailableSpaceInMb(): Int {
            val freeBytesExternal = File(LAppResource.application.getExternalFilesDir(null).toString()).freeSpace
            val freeMb = (freeBytesExternal / (1024 * 1024)).toInt()
//            val totalSize = File(context.getExternalFilesDir(null).toString()).totalSpace
//            val totalMb = (totalSize / (1024 * 1024)).toInt()
            return freeMb
        }

        fun getAvailableRAM(): Long {
            val memoryInfo = ActivityManager.MemoryInfo()
            val activityManager = LAppResource.application.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
            activityManager.getMemoryInfo(memoryInfo)
            val availableMegs = memoryInfo.availMem / 1048576L
            val percentAvail = memoryInfo.availMem / memoryInfo.totalMem
//            Log.d(logTag, "percentAvail $percentAvail")
            return availableMegs
        }

        fun getDownloader(
                folderName: String? = null,
                token: String? = null,
                url: String,
                onDownloadListener: OnDownloadListener
        ): Downloader {
            if (folderName.isNullOrEmpty()) {
                val map = HashMap<String, String>()
                token?.let {
                    map["Authorization"] = it
                }
                return Downloader.Builder(mContext = LAppResource.application, mUrl = url)
                        .header(map)
                        .downloadListener(onDownloadListener)
                        .build()
            } else {
                val path = getFolderPath(folderName = folderName)
                val map = HashMap<String, String>()
                token?.let {
                    map["Authorization"] = it
                }
                return Downloader.Builder(mContext = LAppResource.application, mUrl = url)
                        .downloadDirectory(path)
                        .header(map)
                        .downloadListener(onDownloadListener)
                        .build()
            }
        }

        //return destination file path
        fun unzip(
                file: File
        ): String? {
            try {
                val filePath = file.path
                val destination = "${file.parent}/"
//                LLog.d(logTag, ">>>unzip filePath $filePath")
//                LLog.d(logTag, ">>>unzip destination $destination")
                val inputStream = FileInputStream(filePath)
                val zipStream = ZipInputStream(inputStream)
                var zipEntry: ZipEntry?
                while (zipStream.nextEntry.also {
                            zipEntry = it
                        } != null) {
//                    LLog.d(logTag, "Unzipping " + zipEntry?.name + " at " + destination)
                    zipEntry?.let { ze ->
                        if (ze.isDirectory) {
                            handleDirectory(dir = ze.name, destination = destination)
                        } else {
                            val fileOutputStream = FileOutputStream(destination + "/" + ze.name)
                            val bufferedOutputStream = BufferedOutputStream(fileOutputStream)
                            val buffer = ByteArray(1024)
                            var read: Int
                            while (zipStream.read(buffer).also { read = it } != -1) {
                                bufferedOutputStream.write(buffer, 0, read)
                            }
                            zipStream.closeEntry()
                            bufferedOutputStream.close()
                            fileOutputStream.close()
                        }
                    }
                }
                zipStream.close()
//                LLog.d(logTag, "Unzipping complete, path :  $destination")
                return destination
            } catch (e: java.lang.Exception) {
//                LLog.e(logTag, "Unzipping failed $e")
                e.printStackTrace()
                return null
            }
        }

        private fun handleDirectory(
                dir: String,
                destination: String
        ) {
            val file = File(destination + dir)
//            LLog.d(logTag, "handleDirectory file ${file.path}")
            if (!file.isDirectory) {
//                LLog.d(logTag, "handleDirectory !file.isDirectory")
                val isSuccess = file.mkdirs()
//                LLog.d(logTag, "handleDirectory !file.isDirectory isSuccess $isSuccess")
            } else {
//                LLog.d(logTag, "handleDirectory file.isDirectory")
            }
        }
    }
}
